worker/src/RTC/ActiveSpeakerObserver.cpp

```diff
@@ -0,0 +1,564 @@
+		if (jsonIntervalIt == data.end() || !jsonIntervalIt->is_number())
+			MS_THROW_TYPE_ERROR("missing interval");
+
+		this->interval = jsonIntervalIt->get<int16_t>();
+
+		if (this->interval < 100)
+			this->interval = 100;
+		else if (this->interval > 5000)
+			this->interval = 5000;
+
+		this->periodicTimer = new Timer(this);
+
+		this->periodicTimer->Start(interval, interval);
+	}
+
+	ActiveSpeakerObserver::~ActiveSpeakerObserver()
+	{
+		MS_TRACE();
+
+		delete this->periodicTimer;
+	}
+
+	void ActiveSpeakerObserver::AddProducer(RTC::Producer* producer)
+	{
+		MS_TRACE();
+
+		if (producer->GetKind() != RTC::Media::Kind::AUDIO)
+			MS_THROW_TYPE_ERROR("not an audio Producer");
+
+		if (this->mapProducerSpeaker.find(producer->id) != this->mapProducerSpeaker.end())
+			MS_THROW_ERROR("Producer already in map");
+
+		this->mapProducerSpeaker[producer->id].producer = producer;
+		this->mapProducerSpeaker[producer->id].speaker  = new Speaker();
+	}
+
+	void ActiveSpeakerObserver::RemoveProducer(RTC::Producer* producer)
+	{
+		MS_TRACE();
+
+		auto it = this->mapProducerSpeaker.find(producer->id);
+
+		if (it == this->mapProducerSpeaker.end())
+		{
+			return;
+		}
+
+		if (it->second.speaker != nullptr)
+		{
+			delete it->second.speaker;
+		}
+
+		this->mapProducerSpeaker.erase(producer->id);
+
+		if (producer->id == this->dominantId)
+		{
+			Update();
+		}
+	}
+
+	void ActiveSpeakerObserver::ProducerResumed(RTC::Producer* producer)
+	{
+		MS_TRACE();
+
+		auto it = this->mapProducerSpeaker.find(producer->id);
+
+		if (it != this->mapProducerSpeaker.end())
+		{
+			auto& rtpObserver = it->second;
+
+			rtpObserver.speaker->paused = false;
+		}
+	}
+
+	void ActiveSpeakerObserver::ProducerPaused(RTC::Producer* producer)
+	{
+		MS_TRACE();
+
+		auto it = this->mapProducerSpeaker.find(producer->id);
+
+		if (it != this->mapProducerSpeaker.end())
+		{
+			auto& rtpObserver = it->second;
+
+			rtpObserver.speaker->paused = true;
+		}
+	}
+
+	void ActiveSpeakerObserver::ReceiveRtpPacket(RTC::Producer* producer, RTC::RtpPacket* packet)
+	{
+		MS_TRACE();
+
+		if (IsPaused())
+			return;
+
+		uint8_t volume;
+		bool voice;
+
+		if (!packet->ReadSsrcAudioLevel(volume, voice))
+			return;
+
+		auto it = this->mapProducerSpeaker.find(producer->id);
+
+		if (it != this->mapProducerSpeaker.end())
+		{
+			auto& rtpObserver = it->second;
+			uint64_t now      = DepLibUV::GetTimeMs();
+
+			rtpObserver.speaker->LevelChanged(volume, now);
+		}
+	}
+
+	void ActiveSpeakerObserver::Paused()
+	{
+		MS_TRACE();
+
+		this->periodicTimer->Stop();
+	}
+
+	void ActiveSpeakerObserver::Resumed()
+	{
+		MS_TRACE();
+
+		this->periodicTimer->Restart();
+	}
+
+	void ActiveSpeakerObserver::OnTimer(Timer* timer)
+	{
+		MS_TRACE();
+
+		Update();
+	}
+
+	void ActiveSpeakerObserver::Update()
+	{
+		MS_TRACE();
+
+		uint64_t now              = DepLibUV::GetTimeMs();
+		uint64_t levelIdleTimeout = LevelIdleTimeout - (now - this->lastLevelIdleTime);
+
+		if (levelIdleTimeout <= 0)
```

Comment:
Codady says that this condition doesn't make sense since `levelIdleTimeout` is an unsigned number:

https://app.codacy.com/manual/versatica/mediasoup/pullRequest?bid=24405941&prid=7734463&utm_campaign=new_commit&utm_medium=Email&utm_source=Internal

So, I see two choices here (but I don't know which one is the correct one):

1. Make `levelIdleTimeout` be `int64_t`.
2. Or remove the condition.

Reply:
Given there is a subtraction involved in the calculation, I am going to error on the side of caution, and make this signed.